### DISCUSSION TITLE
资源图像生成

### DISCUSSION THREAD

**Discussion by mark9804** [Arondight/Adachi-BOT#893] [General]
# 图像生成

### png 转 webp

目前有两种操作。

1. [TinyPNG API](https://tinypng.com/developers/reference/nodejs)
2. [cwebp](https://www.npmjs.com/package/webp-converter#user-content-convert-other-image-format-to-webp)

第一种有每人每月 500 张图片的上限，好处是全部扔给它就行了，不用管它后台是怎么做到的

第二种可以选择检查 png 文件头或者显式声明的方式确定需不需要保留透明度图层（alpha 图层）

[Check image transparency](https://stackoverflow.com/a/41302302/14084633)

如果确定不了的话，默认保留 alpha 图层（体积大概会比不保留增加 $\frac{1}{4}$ - $\frac{1}{5}$ 左右）

cwebp 参数：

```shell
cwebp (-noalpha) -m 6 -mt -af -q 95 in.png -o out.webp
```

## 角色

从 https://api.ambr.top/v2/chs/avatar/ 中获取需要的角色，给出一个例子如下：

```json
"10000002": {
    "id": 10000002,
    "rank": 5,
    "name": "神里绫华",
    "element": "Ice",
    "weaponType": "WEAPON_SWORD_ONE_HAND",
    "icon": "UI_AvatarIcon_Ayaka",
    "birthday": [
        9,
        28
    ],
    "release": 1626814800,
    "route": "Kamisato Ayaka"
},
```

https://api.ambr.top/assets/UI/UI_AvatarIcon_Heizo.png

### 角色素材图片

基于 ambr 爬虫，获取素材 ID之后，通过

```
https://api.ambr.top/assets/UI/UI_ItemIcon_${itemID}.png
```

获得

### 角色证件照

根据角色的 `icon` 字段，通过

```
https://api.ambr.top/assets/UI/${icon}.png
```

获得

### 角色立绘

先从内鬼网抓，之后就废弃了

### 角色卡池图片

ambr 没有，只能靠内鬼网的资源助手

### 角色名片

基于 ambr 爬虫，在 

```
RESPONSE.data.other.nameCard.icon
```

里面获得 raw data

例：`icon: "UI_NameCardIcon_Ayaka"`

进行替换找到真实路径

```js
const src =
  "https://api.ambr.top/assets/UI/namecard/" +
  RESPONSE.data.other.nameCard.icon.replace("Icon", "Pic") +
  "_P.png";
```

### 武器素材图片

和角色素材图片一样

### 武器证件照

和角色素材图片一样

### 武器立绘

先手动苟一下，很快就废弃了

### 武器卡池图片

ambr 没有，只能靠内鬼网的资源助手

**Comment by Arondight** [comment]
感谢老大哥，不过图片转换我想尽量本地做 https://github.com/lovell/sharp

**Reply by Arondight** [reply]
https://sharp.pixelplumbing.com/install#chinese-mirror

**Reply by mark9804** [reply]
我上面那个 cwebp 的链接老大哥看过了吗

**Reply by mark9804** [reply]
https://www.npmjs.com/package/webp-converter#user-content-convert-other-image-format-to-webp

**Reply by mark9804** [reply]
看了一下，sharp 会更新一些，上面那个 cwebp 只是单纯的胶水，老大哥如果想用 sharp 的话麻烦先写一个 demo 吧，我平时用的是 cwebp，sharp 那边的转换和压缩参数要试一试才知道哪些是最优的

**Reply by Arondight** [reply]
完全没空，准备月底再做

**Comment by Arondight** [comment]
时装好像只能手动？

**Reply by mark9804** [reply]
我有空看一下，之前都是 fandom 上面抓的，可能时装只能手动了（找不到时装图片的话会 fallback 的）

**Comment by Arondight** [comment]
> ambr 没有，只能靠内鬼网的资源助手

资源助手没这功能吧

**Reply by Arondight** [reply]
看了一下资源助手的使命结束了，内鬼网全部换成 webp 了

**Reply by mark9804** [reply]
内鬼网的 webp 第一质量一般，第二还是裁切过的图，直接用还是有点问题

**Reply by Arondight** [reply]
> 直接用还是有点问题

那需要做什么

**Reply by Arondight** [reply]
我看似乎是一样的

https://genshin.honeyhunterworld.com/img/shougun_052_gacha_card.png

https://genshin.honeyhunterworld.com/img/shougun_052_gacha_card.webp

https://genshin.honeyhunterworld.com/img/i_n12401_gacha_icon.png

https://genshin.honeyhunterworld.com/img/i_n12401_gacha_icon.webp

**Reply by mark9804** [reply]
> 是裁切过的图

武器可以直接用，但是人物的图要检查一下宽度是不是 320px

**Reply by Arondight** [reply]
麻了，角色的能直接用，武器的稀烂

**Reply by mark9804** [reply]
武器的不用管了
我只对角色有要求

**Reply by mark9804** [reply]
> 只对角色有要求

现在记不太清是谁了，当时我记得有些人物的卡池图片周围有空白，被裁掉了，所以那些图横向宽度也不是 320，现在一时半会找不到，总之要假定内鬼网的图不可信

**Reply by Arondight** [reply]
> 武器的不用管了
> 我只对角色有要求

我都裁剪好了

**Comment by Arondight** [comment]
时装证件照是刚需256x256还是1:1即可？

**Reply by mark9804** [reply]
前端只对比例有要求，但是证件照这一类型只可能是 `256×256` 这一种尺寸，如果出现了一个 `256×243` 之类的肯定是被裁剪掉了头上的空白空间

**Reply by Arondight** [reply]
时装的大图以后也不用了吧？

**Reply by mark9804** [reply]
在写了在写了（新建文件夹）

写完就不用了

**Reply by Arondight** [reply]
嗯，之后我把时装的 icon 也换成 webp，大图就不管了
